test/sys/test_socket.rs

```diff
@@ -568,12 +569,142 @@ pub fn test_recv_ipv4pktinfo() {
     thread.join().unwrap();
 }
 
+#[cfg(any(
+    target_os = "freebsd",
+    target_os = "ios",
+    target_os = "macos",
+    target_os = "netbsd",
+    target_os = "openbsd",
+))]
+// qemu doesn't seem to be emulating this correctly in these architectures
+#[cfg_attr(any(
+    target_arch = "mips",
+    target_arch = "mips64",
+    target_arch = "powerpc64",
+), ignore)]
+#[test]
+pub fn test_recvif() {
+    use libc;
+    use nix::ifaddrs::{getifaddrs, InterfaceAddress};
+    use nix::net::if_::*;
+    use nix::sys::socket::sockopt::{Ipv4RecvIf, Ipv4RecvDstAddr};
+    use nix::sys::socket::{bind, AddressFamily, SockFlag, SockType};
+    use nix::sys::socket::{getsockname, setsockopt, socket, SockAddr};
+    use nix::sys::socket::{recvmsg, sendmsg, CmsgSpace, ControlMessage, MsgFlags};
+    use nix::sys::uio::IoVec;
+    use std::io;
+    use std::io::Write;
+    use std::thread;
+
+    fn loopback_v4addr() -> Option<InterfaceAddress> {
+        let addrs = match getifaddrs() {
+            Ok(iter) => iter,
+            Err(e) => {
+                let stdioerr = io::stderr();
+                let mut handle = stdioerr.lock();
+                writeln!(handle, "getifaddrs: {:?}", e).unwrap();
+                return None;
+            },
+        };
+        for ifaddr in addrs {
+            if ifaddr.flags.contains(InterfaceFlags::IFF_LOOPBACK) {
+                match ifaddr.address {
+                    Some(SockAddr::Inet(InetAddr::V4(..))) => {
+                        return Some(ifaddr);
+                    }
+                    _ => continue,
+                }
+            }
+        }
+        None
+    }
+
+    let lo_ifaddr = loopback_v4addr();
+    let (lo_name, lo) = match lo_ifaddr {
+        Some(ifaddr) => (ifaddr.interface_name,
+                         ifaddr.address.expect("Expect IPv4 address on interface")),
+        None => return,
+    };
+    let receive = socket(
+        AddressFamily::Inet,
+        SockType::Datagram,
+        SockFlag::empty(),
+        None,
+    ).expect("receive socket failed");
+    bind(receive, &lo).expect("bind failed");
+    let sa = getsockname(receive).expect("getsockname failed");
+    setsockopt(receive, Ipv4RecvIf, &true).expect("setsockopt IP_RECVIF failed");
+    setsockopt(receive, Ipv4RecvDstAddr, &true).expect("setsockopt IP_RECVDSTADDR failed");
+
+    let thread = thread::spawn(move || {
+        let mut buf = [0u8; 8];
+        let iovec = [IoVec::from_mut_slice(&mut buf)];
+        let mut space = CmsgSpace::<(libc::sockaddr_dl, CmsgSpace<libc::in_addr>)>::new();
+        let msg = recvmsg(
+            receive,
+            &iovec,
+            Some(&mut space),
+            MsgFlags::empty(),
+        ).expect("recvmsg failed");
+        assert!(
+            !msg.flags
+                .intersects(MsgFlags::MSG_TRUNC | MsgFlags::MSG_CTRUNC)
+        );
+
+        for cmsg in msg.cmsgs() {
```

Comment:
There's nothing to ensure that cmsgs were actually received.  Shouldn't you have some kind of assertion to that effect?

Reply:
Your latest updated didn't address this issue.  Do you plan to?

Reply:
Do you just want something like this:
```
assert_eq!(msg.cmsgs().count(), 2, "expected 2 cmsgs");
```
or did you have something different in mind?

Reply:
That statement wouldn't check that you got the right types of cmsgs.  I think you need to check that the cmsgs were of the right types, too.

Reply:
Ok, just added some code that I think covers this. So now, I check to make sure they're are only two and I check to make sure each type is received and it fails if something other than the expected two message types are received.